File: LexiconLMS/Scripts/GenericCompare.js
$.validator.addMethod("genericcompare",
    function(value, element, params) {
        var propelename = params.split(",")[0];
        var operName = params.split(",")[1];
        if (params.length === 0 ||
            value == undefined ||
            value.length === 0 ||
            propelename == undefined ||
            propelename.length === 0 ||
            operName == undefined ||
            operName.length === 0)
            return true;
        var valueOther = $(propelename).val();
        var val1 = (isNaN(value) ? Date.parse(value) : eval(value));
        var val2 = (isNaN(valueOther) ? Date.parse(valueOther) : eval(valueOther));

        if (operName === "GreaterThan")
            return val1 > val2;
        if (operName === "LessThan")
            return val1 < val2;
        if (operName === "GreaterThanOrEqual")
            return val1 >= val2;
        if (operName === "LessThanOrEqual")
            return val1 <= val2;
        if (operName === "Equal")
            return val1 === val2;
        if (operName === "NotEqual")
            return val1 !== val2;
        return true; // Default to true, we can't verify if we don't know how to verify it
    });

$.validator.unobtrusive.adapters.add("genericcompare",
    ["comparetopropertyname", "operatorname"],
    function(options) {
        options.rules["genericcompare"] = "#" +
            options.params.comparetopropertyname +
            "," +
            options.params.operatorname;
        var operstring = "";
        if (options.params.operatorname === "GreaterThan")
            operstring = "greater than";
        if (options.params.operatorname === "LessThan")
            operstring = "greater than or equal to";
        if (options.params.operatorname === "GreaterThanOrEqual")
            operstring = "less than";
        if (options.params.operatorname === "LessThanOrEqual")
            operstring = "less than or equal to";
        if (options.params.operatorname === "Equal")
            operstring = "equal to";
        if (options.params.operatorname === "NotEqual")
            operstring = "not equal to";

        function format(str, obj) {
            return str.replace(/\{\s*([^}\s]+)\s*\}/g, function(m, index, o, s) { return obj[index]; });
        }

        options.messages["genericcompare"] = format(options.message, [operstring]);
        
    });

File: LexiconLMS/Models/GenericCompare.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Web.Mvc;

namespace LexiconLMS.Models
{
    public enum GenericCompareOperator
    {
        GreaterThan,
        GreaterThanOrEqual,
        LessThan,
        LessThanOrEqual,
        Equal,
        NotEqual
    }

    public sealed class GenericCompareAttribute : ValidationAttribute, IClientValidatable
    {
        public string CompareToPropertyName { get; set; }
        public GenericCompareOperator OperatorName { get; set; } = GenericCompareOperator.GreaterThanOrEqual;

        public IEnumerable<ModelClientValidationRule> GetClientValidationRules(ModelMetadata metadata,
            ControllerContext context)
        {
            var errorMessage = FormatErrorMessage(metadata.DisplayName);
            var compareRule = new ModelClientValidationRule
            {
                ErrorMessage = errorMessage,
                ValidationType = "genericcompare"
            };
            compareRule.ValidationParameters.Add("comparetopropertyname", CompareToPropertyName);
            compareRule.ValidationParameters.Add("operatorname", OperatorName.ToString());
            yield return compareRule;
        }

        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
        {
            string operstring;
            switch (OperatorName)
            {
                case GenericCompareOperator.GreaterThan:
                    operstring = "greater than";
                    break;
                case GenericCompareOperator.GreaterThanOrEqual:
                    operstring = "greater than or equal to";
                    break;
                case GenericCompareOperator.LessThan:
                    operstring = "less than";
                    break;
                case GenericCompareOperator.LessThanOrEqual:
                    operstring = "less than or equal to";
                    break;
                case GenericCompareOperator.Equal:
                    operstring = "equal to";
                    break;
                case GenericCompareOperator.NotEqual:
                    operstring = "not equal to";
                    break;
                default:
                    return null; // We can't compare as we don't know what method to use
            }

            var basePropertyInfo = validationContext.ObjectType.GetProperty(CompareToPropertyName);

            var valOther = (IComparable) basePropertyInfo.GetValue(validationContext.ObjectInstance, index: null);
            var valThis = (IComparable) value;

            var compareResult = valThis.CompareTo(valOther);

            if (OperatorName == GenericCompareOperator.GreaterThan && compareResult <= 0 ||
                OperatorName == GenericCompareOperator.GreaterThanOrEqual && compareResult < 0 ||
                OperatorName == GenericCompareOperator.LessThan && compareResult >= 0 ||
                OperatorName == GenericCompareOperator.LessThanOrEqual && compareResult > 0 ||
                OperatorName == GenericCompareOperator.Equal && compareResult == 0 ||
                OperatorName == GenericCompareOperator.NotEqual && compareResult != 0)
            {
                return new ValidationResult(string.Format(ErrorMessage, operstring));
            }
            return null;
        }
    }
}
